doc/arch/008-screen-architecture-v3.md

```diff
@@ -0,0 +1,759 @@
+# ADR 008: Screen Architecture (v3)
+
+## Status
+Accepted (Supersedes [001](./001-screen-controllers.md), [007](./007-screen-architecture-v2.md)) on 2019-10-01
+
+## Context
+Our [current screen architecture](./007-screen-architecture-v2.md) was a step above the [v1 screen architecture](./001-screen-controllers.md). However, it still has problems of its own:
+
+- The `UiStateProducer` is still some sort of a **God Class** since it is reponsible for both business logic (what needs to happen) and effects (changes to the system). This makes testing it cumbersome since the tests are more of integration tests which end up verifying both business logic and implementation details. 
+
+We need a scalable framework to build screens that lets us manage business logic and effects, and which is receptive to change.
+
+## Decision
+
+### Goals
+- Fulfill all the requirements of the [v2 architecture](./007-screen-architecture-v2.md#goals).
+- Separate business logic and effects.
+- Separate tests for business logic and effects so that we can manage them more easily.
+- Have a way to migrate from the v1 architecture to the v3 for existing screens.
+
```

Comment:
We should also add the following:
1. Increase our TAT for feature delivery before and how its impacting the overall project
2. Refactoring code is an increasingly hard problem
3. Onboarding engineers to our old arch required sufficient effort in understanding complexity of RxJava.